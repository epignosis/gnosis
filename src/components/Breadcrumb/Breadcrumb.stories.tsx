import React from "react";
import { Story } from "@storybook/react";
import Breadcrumb, { BreadcrumbProps } from "./Breadcrumb";

export default {
  component: Breadcrumb,
  title: "Components/Breadcrumb",
  argTypes: {
    breadcrumbEl: {
      control: false,
    },
  },
};

export const OneBreadcrumb: Story<BreadcrumbProps> = (args) => {
  return (
    <Breadcrumb {...args}>
      <Breadcrumb.Item>Home</Breadcrumb.Item>
    </Breadcrumb>
  );
};

OneBreadcrumb.args = {};

export const MultipleBreadcrumbs: Story<BreadcrumbProps> = (args) => {
  return (
    <Breadcrumb {...args}>
      <Breadcrumb.Item>
        <a href="#">Home</a>
      </Breadcrumb.Item>
      <Breadcrumb.Item>
        <a href="#">Catalog</a>
      </Breadcrumb.Item>
      <Breadcrumb.Item current> My courses</Breadcrumb.Item>
    </Breadcrumb>
  );
};

MultipleBreadcrumbs.args = {};
