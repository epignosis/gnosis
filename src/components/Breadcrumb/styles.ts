import { css, Theme, SerializedStyles } from "@emotion/react";
import { mq } from "@theme/utils/breakpoints";

export const container = (
  { breadcrumb, typeScaleSizes }: Theme,
  { separator }: { separator: string },
): SerializedStyles => {
  return css`
    background: ${breadcrumb.background};
    color: ${breadcrumb.color};
    padding: 0.875rem 1rem;

    ${mq["lg"]} {
      padding: 1rem 2rem;
    }

    ol,
    li {
      margin: 0;
      padding: 0;
      list-style-type: none;
    }

    ol {
      display: flex;
      flex-wrap: wrap;
    }

    li {
      margin-left: 0.375rem;

      &:first-of-type {
        margin: 0;
      }

      &:after {
        content: " ${separator}";
      }

      &:last-of-type {
        &:after {
          content: "";
        }
      }
    }

    a {
      font-size: ${typeScaleSizes.xs};
    }

    .current {
      font-weight: 700;
      color: ${breadcrumb.color};
      cursor: default;

      &:hover {
        color: ${breadcrumb.color};
        text-decoration: none;
      }
    }
  `;
};
