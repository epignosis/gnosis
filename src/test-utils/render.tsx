import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
