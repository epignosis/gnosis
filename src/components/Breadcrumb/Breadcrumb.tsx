import React, { FC, ReactNode } from "react";
import { createPortal } from "react-dom";
import { SerializedStyles } from "@emotion/react";
import { container } from "./styles";

export type BreadcrumbItemProps = {
  current?: boolean;
};

const Item: FC<BreadcrumbItemProps> = ({ children, current = false }) => {
  return <li className={current ? "breadcrumb-item current" : "breadcrumb-item"}>{children}</li>;
};

export type BreadcrumbProps = React.HTMLAttributes<HTMLElement> & {
  children: ReactNode;
  breadcrumbEl?: Element | null;
  separator?: string;
};

export type BreadcrumbCompoundProps = {
  Item: FC<BreadcrumbItemProps>;
};

const Breadcrumb: FC<BreadcrumbProps> & BreadcrumbCompoundProps = ({
  children,
  breadcrumbEl = null,
  separator = "/",
  ...rest
}) => {
  const breadcrumb = (
    <nav
      css={(theme): SerializedStyles => container(theme, { separator })}
      aria-label="breadcrumbs"
      {...rest}
    >
      <ol>{children}</ol>
    </nav>
  );

  return breadcrumbEl ? createPortal(breadcrumb, breadcrumbEl) : breadcrumb;
};

Breadcrumb.Item = Item;

export default Breadcrumb;
