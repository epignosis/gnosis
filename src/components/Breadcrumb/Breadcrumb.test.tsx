import React from "react";
import { faker } from "@faker-js/faker";
import Breadcrumb from "./Breadcrumb";
import { render, screen } from "@test-utils/render";

const getBreadcrumbProps = () => ({
  item1: faker.unique(faker.lorem.word),
  item2: faker.unique(faker.lorem.word),
  item3: faker.unique(faker.lorem.word),
  item1Url: faker.unique(faker.internet.url),
  item2Url: faker.unique(faker.internet.url),
});

describe("<Breadcrumb> and <Breadcrumb.Item>", () => {
  const { item1, item2, item3, item1Url, item2Url } = getBreadcrumbProps();

  it("renders correctly", () => {
    render(
      <Breadcrumb separator="/">
        <Breadcrumb.Item>
          <a href={item1Url}>{item1}</a>
        </Breadcrumb.Item>
        <Breadcrumb.Item>
          <a href={item2Url}>{item2}</a>
        </Breadcrumb.Item>
        <Breadcrumb.Item current> {item3}</Breadcrumb.Item>
      </Breadcrumb>,
    );

    const firstBreadcrumbLink = screen.getByText(item1);
    const secondBreadcrumbLink = screen.getByText(item2);
    const currentBreadcrumb = screen.getByText(item3);

    expect(firstBreadcrumbLink).toHaveTextContent(item1);
    expect(secondBreadcrumbLink).toHaveTextContent(item2);
    expect(currentBreadcrumb).toHaveTextContent(item3);
    expect(firstBreadcrumbLink).toHaveAttribute("href", item1Url);
    expect(secondBreadcrumbLink).toHaveAttribute("href", item2Url);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Breadcrumb id="main-nav" className="navigation" separator="/">
        <Breadcrumb.Item>
          <a href="#/home">Home</a>
        </Breadcrumb.Item>
        <Breadcrumb.Item>
          <a href="#/my-courses">My courses</a>
        </Breadcrumb.Item>
        <Breadcrumb.Item current> My super course</Breadcrumb.Item>
      </Breadcrumb>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders correctly in portal", () => {
    const breadcrumbContainer = document.createElement("div");
    breadcrumbContainer.setAttribute("id", "breadcrumb");
    document.body.appendChild(breadcrumbContainer);

    render(
      <Breadcrumb breadcrumbEl={breadcrumbContainer} separator="/">
        <Breadcrumb.Item>
          <a href="#/home">Home</a>
        </Breadcrumb.Item>
        <Breadcrumb.Item>
          <a href="#/my-courses">My courses</a>
        </Breadcrumb.Item>
        <Breadcrumb.Item current> My super course</Breadcrumb.Item>
      </Breadcrumb>,
    );
    const breadcrumbs = screen.getByLabelText("breadcrumbs");

    expect(breadcrumbs).toMatchSnapshot();
  });
});
