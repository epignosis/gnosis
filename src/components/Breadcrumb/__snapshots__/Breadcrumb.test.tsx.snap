// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumb> and <Breadcrumb.Item> matches snapshot 1`] = `
<div>
  <nav
    aria-label="breadcrumbs"
    class="navigation css-1vr3ddk-container-container"
    id="main-nav"
  >
    <ol>
      <li
        class="breadcrumb-item"
      >
        <a
          href="#/home"
        >
          Home
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <a
          href="#/my-courses"
        >
          My courses
        </a>
      </li>
      <li
        class="breadcrumb-item current"
      >
         My super course
      </li>
    </ol>
  </nav>
</div>
`;

exports[`<Breadcrumb> and <Breadcrumb.Item> renders correctly in portal 1`] = `
<nav
  aria-label="breadcrumbs"
  class="css-1vr3ddk-container-container"
>
  <ol>
    <li
      class="breadcrumb-item"
    >
      <a
        href="#/home"
      >
        Home
      </a>
    </li>
    <li
      class="breadcrumb-item"
    >
      <a
        href="#/my-courses"
      >
        My courses
      </a>
    </li>
    <li
      class="breadcrumb-item current"
    >
       My super course
    </li>
  </ol>
</nav>
`;
